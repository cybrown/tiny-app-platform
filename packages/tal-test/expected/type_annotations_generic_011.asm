main()
  entry:
    FunctionRef       name: map_0
    DeclareLocal      name: map, mutable: false, hasInitialValue: true
    Pop               inBlock: false
    FunctionRef       name: number_to_string_1
    DeclareLocal      name: number_to_string, mutable: false, hasInitialValue: true
    Pop               inBlock: false
    Literal           0
    MakeArray
    Literal           "value"
    Literal           true
    Literal           1
    MakeArray
    Literal           "mapper"
    FunctionRef       name: func_2
    Literal           2
    MakeRecord
    Local             map
    Call
map_0(value, mapper)
  entry:
    Literal           0
    Local             value
    Index
    Literal           1
    MakeArray
    Literal           0
    MakeRecord
    Local             mapper
    Call
    Literal           1
    MakeArray
    MakeArrayForBlock count: 1
number_to_string_1(num)
  entry:
    Literal           ""
    MakeArrayForBlock count: 1
func_2(item)
  entry:
    Literal           "Plus 1: "
    Local             item
    Literal           1
    Intrinsic         INTRINSIC_ADD
    Literal           1
    MakeArray
    Literal           0
    MakeRecord
    Local             number_to_string
    Call
    Intrinsic         INTRINSIC_ADD
