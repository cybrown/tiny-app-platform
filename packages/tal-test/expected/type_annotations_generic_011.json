[
  {
    "kind": "DeclareLocal",
    "mutable": false,
    "name": "map",
    "value": {
      "kind": "Function",
      "body": {
        "kind": "Block",
        "children": [
          {
            "kind": "Array",
            "value": [
              {
                "kind": "Call",
                "args": [
                  {
                    "kind": "PositionalArgument",
                    "value": {
                      "kind": "Index",
                      "index": 0,
                      "value": {
                        "kind": "Local",
                        "name": "value"
                      }
                    }
                  }
                ],
                "value": {
                  "kind": "Local",
                  "name": "mapper"
                }
              }
            ]
          }
        ]
      },
      "parameters": [
        {
          "name": {
            "kind": "Identifier",
            "name": "value"
          },
          "type": {
            "kind": "array",
            "item": {
              "kind": "named",
              "name": "T"
            }
          }
        },
        {
          "name": {
            "kind": "Identifier",
            "name": "mapper"
          },
          "type": {
            "kind": "function",
            "parameters": [
              {
                "name": {
                  "kind": "Identifier",
                  "name": "item"
                },
                "type": {
                  "kind": "named",
                  "name": "T"
                }
              }
            ],
            "returnType": {
              "kind": "named",
              "name": "U"
            }
          }
        }
      ],
      "returnType": {
        "kind": "array",
        "item": {
          "kind": "named",
          "name": "U"
        }
      },
      "genericParameters": [
        {
          "name": "T"
        },
        {
          "name": "U"
        }
      ]
    }
  },
  {
    "kind": "DeclareLocal",
    "mutable": false,
    "name": "number_to_string",
    "value": {
      "kind": "Function",
      "body": {
        "kind": "Block",
        "children": [
          ""
        ]
      },
      "parameters": [
        {
          "name": {
            "kind": "Identifier",
            "name": "num"
          },
          "type": {
            "kind": "named",
            "name": "number"
          }
        }
      ],
      "returnType": {
        "kind": "named",
        "name": "string"
      }
    }
  },
  {
    "kind": "Call",
    "args": [
      {
        "kind": "NamedArgument",
        "name": {
          "kind": "Identifier",
          "name": "value"
        },
        "value": {
          "kind": "Array",
          "value": [
            true
          ]
        }
      },
      {
        "kind": "NamedArgument",
        "name": {
          "kind": "Identifier",
          "name": "mapper"
        },
        "value": {
          "kind": "Function",
          "body": {
            "kind": "BinaryOperator",
            "operator": "+",
            "right": {
              "kind": "Call",
              "args": [
                {
                  "kind": "PositionalArgument",
                  "value": {
                    "kind": "BinaryOperator",
                    "operator": "+",
                    "right": 1,
                    "left": {
                      "kind": "Local",
                      "name": "item"
                    }
                  }
                }
              ],
              "value": {
                "kind": "Local",
                "name": "number_to_string"
              }
            },
            "left": "Plus 1: "
          },
          "parameters": [
            {
              "name": {
                "kind": "Identifier",
                "name": "item"
              }
            }
          ]
        }
      }
    ],
    "value": {
      "kind": "Local",
      "name": "map"
    }
  }
]